### PR TITLE
Multiprocess

### PR DESCRIPTION
**Pull request by bonham79**
Multiprocessor fix for training loop. This should work on multi GPU setup. I confirmed on a 1080 + 3080 (yes, I'm also amazed they're compatible). This was done by moving the inner functions from `get_loss_func` to being top level functions. (Required for pickle.)  Also made namespace tweaks to be more PTL agreeable.

Thing to note: I moved transpose behavior into training and validation loops. See if this is agreeable or we'd prefer to do the change at model level.  (Issue discussed in https://github.com/CUNY-CL/yoyodyne/pull/44)

### CODE REVIEW

File: yoyodyne/models/transducer.py
Comment (kylebgorman) [inline yoyodyne/models/transducer.py:512]: May I suggest: 

```python
_, loss = self(batch)
```

would be the clearest possible, since it is a surprise on line 514 to learn that something you call predictions has the loss in its second element.

**Review by kylebgorman — APPROVED** [review]
I reviewed this by reading the last (of 11) commits, bc4be28fd40548d5011656bc4c815aa6d71dfe52. This seems like a straightforward improvement. (I also like that we have one fewer subdefinition, just in general.) LGTM.

Can I ask about the name change from `self.loss_func` to just `self.loss`? I think the old one was better but there may be some justification I'm missing.

**Comment by bonham79** [comment]
> Can I ask about the name change from `self.loss_func` to just `self.loss`? I think the old one was better but there may be some justification I'm missing.

Self.loss is more common in lightningese. That's all.

**Comment by bonham79** [comment]
> Can I ask about the name change from `self.loss_func` to just `self.loss`? I think the old one was better but there may be some justification I'm missing.

Self.loss is more common in lightningese. That's all.

**Comment by bonham79** [comment]
> I reviewed this by reading the last (of 11) commits, [bc4be28](https://github.com/CUNY-CL/yoyodyne/commit/bc4be28fd40548d5011656bc4c815aa6d71dfe52). This seems like a straightforward improvement. (I also like that we have one fewer subdefinition, just in general.) LGTM.
> 
Sorry bout that!  There was a version of 'black' in my 'bin' that decided to take supremacy over my env....

**Comment by kylebgorman** [comment]
So this is fine but I'm going to wait until the earlier PR is approved by @Adamits before I merge.

**Review by Adamits — APPROVED** [review]
I forgot about the inner function multiprocessing issue! Good catch!

I think moving the transposes is an improvement. I can't really evaluate if the transpose changes are correct without running this -- but it sounds like you tested all of this in both training and prediction?

I also agree with Kyle that I like `loss_func` more, is there a doc anywhere demonstrating that `loss` should be used for the name of the function? Or is it just a general thing you've picked up from working on other PTL codebases? Either way I am fine with it if you are pretty sure it's justified :D.

LGTM

**Comment by kylebgorman** [comment]
okay @bonham79 should we merge this?

(I may revert `loss_func` later.)

**Comment by bonham79** [comment]
> I forgot about the inner function multiprocessing issue! Good catch!
> 
> I think moving the transposes is an improvement. I can't really evaluate if the transpose changes are correct without running this -- but it sounds like you tested all of this in both training and prediction?
> 

I did some quick testing on the english dataset and there were no flags set off by predict. 

> I also agree with Kyle that I like `loss_func` more, is there a doc anywhere demonstrating that `loss` should be used for the name of the function? Or is it just a general thing you've picked up from working on other PTL codebases? Either way I am fine with it if you are pretty sure it's justified :D.

Just something I picked up from other codebases. I don't think it violates any style rules, just strikes me as odd. I'm fine with reversion if the tribunal is 2-to-1.

**Comment by bonham79** [comment]
> okay @bonham79 should we merge this?
> 
> (I may revert `loss_func` later.)

@Adamits Are there any issues in the other PR that aren't caught by this one?

**Comment by Adamits** [comment]
> > okay @bonham79 should we merge this?
> > (I may revert `loss_func` later.)
> 
> @Adamits Are there any issues in the other PR that aren't caught by this one?

I don't think so, thanks!

**Comment by bonham79** [comment]
@kylebgorman Commit away!

**Pull request merged**
